nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,730 @@
+        streamlines_gen = iter([])
+        if self._streamlines is not None:
+            streamlines_gen = self._streamlines()
+        elif self._data is not None:
+            streamlines_gen = (t.streamline for t in self._data())
+
+        # Check if we need to apply an affine.
+        if not np.allclose(self._affine_to_apply, np.eye(4)):
+            def _apply_affine():
+                for s in streamlines_gen:
+                    yield apply_affine(self._affine_to_apply, s)
+
+            return _apply_affine()
+
+        return streamlines_gen
+
+    def _set_streamlines(self, value):
+        if value is not None and not callable(value):
+            raise TypeError("`streamlines` must be a generator function.")
+
+        self._streamlines = value
+
+    @property
+    def data_per_streamline(self):
+        return self._data_per_streamline
+
+    @data_per_streamline.setter
+    def data_per_streamline(self, value):
+        self._data_per_streamline = LazyDict(value)
+
+    @property
+    def data_per_point(self):
+        return self._data_per_point
+
+    @data_per_point.setter
+    def data_per_point(self, value):
+        self._data_per_point = LazyDict(value)
+
+    @property
+    def data(self):
+        if self._data is not None:
+            return self._data()
+
+        def _gen_data():
+            data_per_streamline_generators = {}
+            for k, v in self.data_per_streamline.items():
+                data_per_streamline_generators[k] = iter(v)
+
+            data_per_point_generators = {}
+            for k, v in self.data_per_point.items():
+                data_per_point_generators[k] = iter(v)
+
+            for s in self.streamlines:
+                data_for_streamline = {}
+                for k, v in data_per_streamline_generators.items():
+                    data_for_streamline[k] = next(v)
+
+                data_for_points = {}
+                for k, v in data_per_point_generators.items():
+                    data_for_points[k] = next(v)
+
+                yield TractogramItem(s, data_for_streamline, data_for_points)
+
+        return _gen_data()
+
+    def __getitem__(self, idx):
+        raise NotImplementedError('LazyTractogram does not support indexing.')
+
+    def __iter__(self):
+        count = 0
+        for tractogram_item in self.data:
+            yield tractogram_item
+            count += 1
+
+        # Keep how many streamlines there are in this tractogram.
+        self._nb_streamlines = count
+
+    def __len__(self):
+        # Check if we know how many streamlines there are.
+        if self._nb_streamlines is None:
+            warn("Number of streamlines will be determined manually by looping"
+                 " through the streamlines. If you know the actual number of"
+                 " streamlines, you might want to set it beforehand via"
+                 " `self.header.nb_streamlines`."
+                 " Note this will consume any generators used to create this"
```

Comment:
No, it can. Since `LazyTractogram` objects are initialized with generator functions (i.e. functions that returns a generator), they can be re-iterated over as many time as you like.

This is a trace of the old API that accepted generators instead of generator functions. I will remove that part of the message. Good catch.